src/Data/Nat/Binary/DivMod.agda

```diff
@@ -0,0 +1,189 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Division with remainder on binary represented naturals.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Nat.Binary.DivMod where
+
+open import Data.Nat.Binary.Base
+open import Data.Nat.Binary.Induction
+open import Data.Nat.Binary.Properties
+open import Data.Nat.Binary.Subtraction
+open import Function using (_∘_)
+open import Relation.Binary using (Tri)
+import Relation.Binary.Reasoning.Base.Triple as InequalityReasoning
+open import Relation.Binary.PropositionalEquality as P using
+  (_≡_; _≢_; refl; sym; cong; resp₂)
+open import Induction.WellFounded using (Acc)
+open Acc using (acc)
+open InequalityReasoning {A = ℕᵇ} {_≈_ = _≡_} ≤-isPreorder <-trans (resp₂ _<_)
+                         <⇒≤ <-≤-trans ≤-<-trans
+open ℕᵇ
+open Tri
+
+------------------------------------------------------------------------------
+-- DivMod  is the definition of what is division with remainder,
+-- and it is the result type for division with remainder. It includes certain
+-- two property witnesses for the quotient and remainder.
+--
+-- Note that  divisor ≢ 0ᵇ  can be derived from  res : DivMod _ divisor.
+------------------------------------------------------------------------------
+
+record DivMod (dividend divisor : ℕᵇ) : Set where  -- result of division with remainder
+  constructor divModᶜ
+  field
+    quotient    : ℕᵇ
+    remainder   : ℕᵇ
+    equality    : dividend ≡ remainder + quotient * divisor
+    rem<divisor : remainder < divisor
+
+------------------------------------------------------------------------------
+-- Evaluating division with remainder.
+-- Illustration:  divMod 2a b  for b ≢ 0  is found by reducing it to division of
+-- a  by  b  as follows.
+-- By recursion, let      a = qb + r,    r < b.
+-- Then  2a = 2qb + 2r,   r ≤ 2r < 2b,   r < b  < 2b.
+-- If  2r < b  then the result is (2q , 2r),
+-- otherwise                      (2q+1, 2r - b).
+-- In the latter case it is needed to prove  2r - b < b.
+-- This is equivalent to  2r < 2b  ~  r < b.
+------------------------------------------------------------------------------
+
+divMod : (a b : ℕᵇ) → b ≢ 0ᵇ → DivMod a b
+divMod a b b≢0 =  dm a (<-wellFounded a)
```

Comment:
@mechvel this approach is not practical. The fact that this function takes takes 150 lines to define vs the 6 lines it takes to define `div` and `mod` for `ℕ` should be a strong indication of this. However the main problem is that the values can't be calculated independently from the proofs, and makes using this definition extremely difficult. For example the goal in the first hole in the proof below is is **1552** lines.
```
test : ∀ a → quot a 1ᵇ (λ()) ≡ a
test zero     = refl
test 2[1+ a ] = {!!}
test 1+[2 a ] = {!!}
```
The way to go about this is:
1. define functions calculating only the values
2. prove equivalence to division and remainder for `ℕ`
3. transfer all the proofs across using the equivalence.

I know you have concerns about the performance of step 3, but it's definitely possible to avoid evaluating the proofs if you read up about irrelevance and take a little care about how you define your functions.

Reply:
Well, let me try.